### ISSUE TITLE
Nullable<T>.Value following Nullable<T>.HasValue can be optimized

### ISSUE THREAD

**Issue by nxtn** [dotnet/runtime#13699]
The `hasValue` check inside the `Nullable<T>.Value` getter is unnecessary since `hasValue` is readonly.

```cs
using System.Runtime.CompilerServices;

class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int? GetNullable() => null;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int DoSomething(int index) => index;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int DoSomethingElse(int index) => index;

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    static int Main()
    {
        int? i = GetNullable();
        return i.HasValue ? DoSomething(i.Value) : DoSomethingElse(i.Value);
    }
}
```

```asm
sub     rsp,28h
xor     eax,eax
mov     [rsp+20h],rax

call    Program.GetNullable()
mov     [rsp+20h],rax
cmp     byte ptr [rsp+20h],0
jne     short LBL_0

cmp     byte ptr [rsp+20h],0 ; <=
je      short LBL_1
mov     ecx,[rsp+24h]
call    Program.DoSomethingElse(Int32)

nop
add     rsp,28h
ret

LBL_0:
cmp     byte ptr [rsp+20h],0 ; <=
je      short LBL_2
mov     ecx,[rsp+24h]
call    Program.DoSomething(Int32)

nop
add     rsp,28h
ret

; Do we need two identical code blocks?

LBL_1:
call    System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_NoValue()
int     3

LBL_2:
call    System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_NoValue()
int     3
```

category:cq
theme:optimization
skill-level:intermediate
cost:medium

**Comment by xtqqczze** [comment]
`n ?? -1` can be optimised too:

```csharp
    public int CoalesceExpression(int? n) {
        return n ?? -1;
    }
    public int Optimised(int? n) {
        if (n.HasValue)
            return n.GetValueOrDefault();
        return -1;
    }
```

```
; Core CLR 5.0.721.25508 on amd64

C..ctor()
    L0000: ret

C.CoalesceExpression(System.Nullable`1<Int32>)
    L0000: mov [rsp+0x10], rdx
    L0005: lea rax, [rsp+0x10]
    L000a: movzx edx, byte ptr [rax]
    L000d: mov eax, [rax+4]
    L0010: test dl, dl
    L0012: jne short L001a
    L0014: mov eax, 0xffffffff
    L0019: ret
    L001a: ret

C.Optimised(System.Nullable`1<Int32>)
    L0000: mov [rsp+0x10], rdx
    L0005: cmp byte ptr [rsp+0x10], 0
    L000a: je short L0011
    L000c: mov eax, [rsp+0x14]
    L0010: ret
    L0011: mov eax, 0xffffffff
    L0016: ret
```

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYQFgAoAbyfPcooEsA7DOiNgA2MXGBgBRBAAdYuXFwg8AFLwwB+cjwCU5Vow6HKAdi3l1muAEYA3Gw4Bfe+zLk15APLSMXfF1wwACaqfJo6es5GXABm5Mo8AHQAEngAasIArjDakUYcxKaJAOIwGOlCWR5QACIw0dgZQhjK2nYGeSbk1m2GTowOQA==

**Comment by xtqqczze** [comment]
@stephentoub This is now optimised by the jit in 6.0?

**Comment by stephentoub** [comment]
```C#
    [Benchmark]
    public int GetValue() => GetValue(42);

    [MethodImpl(MethodImplOptions.NoInlining)]
    private int GetValue(int? n)
    {
        if (n.HasValue) return n.Value;
        return -1;
    }
```

.NET 5:
```assembly
; Program.GetValue(System.Nullable`1<Int32>)
       sub       rsp,28
       mov       [rsp+38],rdx
       movzx     eax,byte ptr [rsp+38]
       test      eax,eax
       je        short M01_L00
       test      eax,eax
       je        short M01_L01
       mov       eax,[rsp+3C]
       add       rsp,28
       ret
M01_L00:
       mov       eax,0FFFFFFFF
       add       rsp,28
       ret
M01_L01:
       call      System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_NoValue()
       int       3
; Total bytes of code 47
```

.NET 6:
```assembly
; Program.GetValue(System.Nullable`1<Int32>)
       mov       [rsp+10],rdx
       cmp       byte ptr [rsp+10],0
       je        short M01_L00
       mov       eax,[rsp+14]
       ret
M01_L00:
       mov       eax,0FFFFFFFF
       ret
; Total bytes of code 23
```

**Comment by xtqqczze** [comment]
According to Stephen in his blog post [Performance Improvements in .NET 6](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/), this issue was fixed by #43811 and #46237.
